solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java

```diff
@@ -394,26 +393,26 @@ public void onHeaders(Response response) {
               assert ObjectReleaseTracker.track(is);
               try {
                 NamedList<Object> body = processErrorsAndResponse(solrRequest, parser, response, is);
-                asyncListener.onSuccess(body);
-              } catch (RemoteSolrException e) {
-                if (SolrException.getRootCause(e) != CANCELLED_EXCEPTION) {
-                  asyncListener.onFailure(e);
-                }
-              } catch (SolrServerException e) {
-                asyncListener.onFailure(e);
+                future.complete(body);
+              } catch (RemoteSolrException | SolrServerException e) {
+                future.completeExceptionally(e);
               }
             });
           }
 
           @Override
           public void onFailure(Response response, Throwable failure) {
             super.onFailure(response, failure);
-            if (failure != CANCELLED_EXCEPTION) {
-              asyncListener.onFailure(new SolrServerException(failure.getMessage(), failure));
-            }
+            future.completeExceptionally(new SolrServerException(failure.getMessage(), failure));
           }
         });
-    return () -> req.abort(CANCELLED_EXCEPTION);
+    future.exceptionally((error) -> {
+      if (error instanceof CancellationException) {
+        req.abort(new Exception());
```

Comment:
I feel the way cancellation is implemented is little bit wrong. When we abort a request the given exception will be called on listeners, in that case we don't want to cascade that exception further. By doing this way, the `new Exception()` will be throw to upper level.

Reply:
Although the exception would be called by listeners, wouldn't the `future.completeExceptionally` that is used to handle the exception be a noop as the future has already been cancelled?

Reply:
I'm a little curious about the code being referenced here:
* Why do an instanceof check on "error" at all; why not simply always req.abort(error)?
* Why create a new Exception when we already have a suitable Throwable type in variable "error"?

Reply:
> Although the exception would be called by listeners, wouldn't the future.completeExceptionally that is used to handle the exception be a noop as the future has already been cancelled?

I see your point, but I in this case from CompletableFuture perspective, it get called with `cancel()` first then `completeExceptionally` will that mess up with internal state of the future instance?